test/e2e/kubedns/kubedns.go

```diff
@@ -55,4 +59,69 @@ var _ = Describe("kube-dns", func() {
 			kubeDNS.Stop()
 		})
 	})
+
+	It("should forward missed ptr lookup to upstream server", func() {
+		By("Setting up environment without upstream server")
+		kubeDNS := &e2edns.KubeDNS{}
+		fr := e2e.GetFramework()
+		configDir := fr.Options.WorkDir + "/kube-dns-config"
+
+		if _, err := os.Stat(configDir); err == nil {
+			os.RemoveAll(configDir)
+		}
+		if err := os.MkdirAll(configDir, 0744); err != nil {
+			panic(err)
+		}
+
+		fr.Docker.Pull(fr.Options.DnsmasqPtrImage)
+		dnsmasq := fr.Docker.Run(
+			"-d",
+			"-p=10054:53/tcp",
+			"-p=10054:53/udp",
+			"--cap-add=NET_ADMIN",
+			fr.Options.DnsmasqPtrImage)
+
+		kubeDNS.Start("kube-dns-ptrfwd", "-v=4", "--config-dir="+configDir)
+
+		By("Get answer without upstream server")
+		om.Eventually(func() error {
+			return doPtrQuery(kubeDNS)
+		}, 1*time.Minute).ShouldNot(om.Succeed())
+
+		By("Config upstream nameserver")
+		if err := ioutil.WriteFile(configDir+"/upstreamNameservers", []byte("[\"127.0.0.1:10054\"]"), 0744); err != nil {
+			panic(err)
+		}
+
+		By("Get answer from upstream server")
+		om.Eventually(func() error {
+			return doPtrQuery(kubeDNS)
+		}, 1*time.Minute).Should(om.Succeed())
+
+		By("Clean up test environment")
+		kubeDNS.Stop()
+		if dnsmasq != "" {
+			fr.Docker.Kill(dnsmasq)
+		}
+
+		if _, err := os.Stat(configDir); err == nil {
+			os.RemoveAll(configDir)
+		}
+	})
 })
+
+func doPtrQuery(kubeDNS *e2edns.KubeDNS) error {
+	time.Sleep(1 * time.Second)
+	names, err := kubeDNS.Query("4.3.2.1.in-addr.arpa.", dns.TypePTR)
```

Comment:
You can use an IP from the testbed range to guarantee no one will reply. `1.2.3.4` may be someone's address.

192.0.2.0/24

https://en.wikipedia.org/wiki/Reserved_IP_addresses

Reply:
IP changed. Now using numb upstream server when test starts and then switch to a upstream server with the desired PTR record.